### DISCUSSION TITLE
FPS alway set to 30 every time IOS is in battery Save Mode (yellow color in  battery)

### DISCUSSION THREAD

**Discussion by sonht-0555** [thenick775/gbajs3#87] [Q&A] [answered]
In savemode, the fps is always set to 30. I don't know how to fix it to 60fps like https://gba.44670.org/. This emulator also runs on Wasm and uses VBA as its core. I don't understand how it doesn't drop fps to 30 when in IOS savemode.

**Comment by sonht-0555** [comment]
Even if I use Module.setMainLoopTiming(0, 1) it doesn't prevent the fps from dropping to 30 and making the game slow.

**Reply by sonht-0555** [reply]
I tried asking a simulator programmer about this. I want to fix it like on the message. Can your core be edited as above?

**Reply by sonht-0555** [reply]
https://github.com/EmulatorJS/RetroArch/blob/a6bd6ff8a6db79b1afeced815eea657998593cd3/gfx/drivers_context/emscriptenegl_ctx.c#L49

**Reply by sonht-0555** [reply]
![image](https://github.com/thenick775/gbajs3/assets/26568422/d4c46d79-f00d-43c8-8e2f-7ea30dc7af84)

**Reply by thenick775** [reply]
> ![image](https://github.com/thenick775/gbajs3/assets/26568422/d4c46d79-f00d-43c8-8e2f-7ea30dc7af84)

Currently fast forward uses setTimeout if you look at the emscripten docs for setting the main loop timing.

Both suffer from the same problem on iOS, as when power saver mode is active, it throttles the cpu.

I haven't worked out a good way around this yet, but feel free to take a stab at it!

**Reply by thenick775** [reply]
I'll see how setimmediate plays with power saver, it's not recommended in the [emscripten docs](https://emscripten.org/docs/api_reference/emscripten.h.html#c.emscripten_set_main_loop_timing), but we'll see how it performs.

Edit: `setImmediate` is deprecated, and is marked non-standard. I most likely won't implement this method, but may experiment, as this is definitely an issue I want to solve.

**Reply by sonht-0555** [reply]
https://demo.emulatorjs.org
Try this simulator?, fast forward is bypassing IOS savemode. I think they fixed that. I don't know how to investigate

**Reply by sonht-0555** [reply]
I also tried changing the mode to setimmediate but our emulator core cannot pass the ios savemode. It seems like it's not the right issue

**Reply by sonht-0555** [reply]
and I also see skyemu solving that problem. I think we will find out the reason soon

**Comment by thenick775** [comment]
@sonht0555 would you be able to help confirm a few things?

I'm now building the mgba binary with pthread support, enabling `sharedArrayBuffer`.

I still need to do some work to make sure things are functioning as expected, but here's some details of what I've been playing around with:

- this require coop/coep headers, so you might have to refresh a few times, or install to the home screen with the service worker
- it seems like after compiling with pthreads, that `setInterval` can move a lot faster than RAF, and seems unaffected by OS level throttling
- this kind of breaks GitHub pages deploys, so I added a temporary service worker addition that will set coop/coep headers on subsequent requests (hence the response above of having to refresh a few times depending on you browser)

What you should see is that `fast forward` mode is **not** throttled if everything goes correctly. I can potentially make this work in regular speed mode as well under low power mode replacing RAF with `setInterval`.

There's a lot left to be done to get this out in a way that it will be fully usable, but this is a good first experiment.

Here's the [link](https://thenick775.github.io/gbajs3/) to my staging deploy based on this [branch](https://github.com/thenick775/gbajs3/compare/master...thenick775:gbajs3:test-wasm-updates) if you want to play around as I try and fix some of the problems listed above!

I'd recommend installing this and having all req go through the service worker for these initial tests.

**Reply by thenick775** [reply]
Here's a demo of this in its current implementation:


https://github.com/thenick775/gbajs3/assets/40526638/9e97a263-7920-4976-9d41-d37fa8e47bfc

**Reply by thenick775** [reply]
Frameskipping with vsync seems to work the best for regular speed.

I'll have to think about how to handle fast forward a bit more, as with timeouts it's currently still throttled after going into the background.

**Comment by sonht-0555** [comment]
I am using ip13 and ios 17.5.1.   There is some problem that makes the emulator not run after I select the ROM

**Reply by sonht-0555** [reply]
https://github.com/thenick775/gbajs3/assets/26568422/18db5425-5c67-4319-afe7-086951e04cda

**Comment by sonht-0555** [comment]
@thenick775 
> this require coop/coep headers, so you might have to refresh a few times, or install to the home screen with the service worker

Regarding service workers, I always want users to put my application on the home screen, leaving it running in safari will encounter some problems such as not being able to save the game after 7 days.

> it seems like after compiling with pthreads, that setInterval can move a lot faster than RAF, and seems unaffected by OS level throttling

As I mentioned in the above messages, other programmers use setInterval to make the simulator run faster in Save_model mode.  But it still has a small problem: the sound is still distorted, but it's not distorted at 30fps but about 50-55fps.  I wonder if you have encountered that?. 

 Recently I discovered that Sky is based on the core idea and is a brother of NBA (NanoBoyAdvance).  I found a programmer who tried recompiling NBA to run on Wasm.  It passes through save_battery and runs very stable on iOS (no overheating like mGBA core) this is it [nealmm.github.io/web-gba-emu/](url) NBA is considered lightweight -  Accurate and mGBA is rated for good feedback - many functions.

> Here's the [link](https://thenick775.github.io/gbajs3/) to my staging deploy based on this [branch](https://github.com/thenick775/gbajs3/compare/master...thenick775:gbajs3:test-wasm-updates) if you want to play around as I try and fix some of the problems listed above!
> 
> I'd recommend installing this and having all req go through the service worker for these initial tests.

You mean I will play the game via home screen shortcut (bring your emulator to the home screen)

**Comment by sonht-0555** [comment]
I hope this idea can be implemented. If possible, it will make the gaming process on this core much longer.

**Reply by thenick775** [reply]
@sonht0555 could you give this a test?

Here is the [link](https://thenick775.github.io/gbajs3/) to the latest deploy, I'd open it in a private tab to make sure you don't have a cached build.

I've fixed a few things, and am trying a completely different approach with only request animation frame as recommended by the emscripten docs!

Fixed a few of the issues above (including the rom loading issue), please let me know if this solves the fps slowdown in low power mode, or if you notice any additional problems!

**Comment by sonht-0555** [comment]
@thenick775  I tested it, and it's a good signal, there are a few points that need to be corrected if you want it to be the best
 - Even though it runs at the correct speed, it seems to stutter.  You can clearly see when I increase to 4x-5x runs
 - The sound is also a bit jerky.  It could be due to intentionally forcing the emulator to run at its normal speed. 
 
 Overall, it's very good, to fix the above two things it will be almost the same as the SkyEmu or Nba core.

**Comment by sonht-0555** [comment]
@thenick775 
https://github.com/thenick775/gbajs3/assets/26568422/8f5d9c81-308b-400c-8e36-75a47f1ec86b

**Reply by thenick775** [reply]
Interesting, I was seeing the same thing for a little bit.

I uninstalled and reinstalled the app to my home screen and that seems to have resolved any stuttering I was seeing.

At 4-5x frame rates some for drops are expected, as really we're going up to potentially 300fps (simulated).

I'll look around a bit more just just to be sure 😎

**Reply by thenick775** [reply]
Are you still seeing stutter at the 1x fast forward (original speed) in low power mode?

**Reply by sonht-0555** [reply]
@thenick775 
> Are you still seeing stutter at the 1x fast forward (original speed) in low power mode?

If you pay attention and still see a bit of lag, this is difficult to fix because as I said above.  You can verify it when looking around at the moving character, the more fps it increases > 60, the more clearly it will be seen.
 Looking forward to the upcoming test versions ^_^

**Reply by sonht-0555** [reply]
@thenick775 By the way, I just tried opening the app again under the system tray, it was frozen, all elements on the screen.  I tried touching, but no response.  You can check to see if the same thing happens :/

**Reply by sonht-0555** [reply]
To reproduce it more easily, open your emulator, play for about 30 seconds at 4x speed.  Then open a heavy game, play that game for 30 seconds and then come back, you will get the error again.

**Reply by thenick775** [reply]
@sonht0555 can you check out the latest staging deploy?

[Link](https://thenick775.github.io/gbajs3/)

This should have resolved the following issues:
- emulator delta time including time in the background (caused the hang you saw)
   - I plan to do a bit more in the core itself to prevent this, but this is a good first pass POC
- performance
   - I've made some adjustments to how app state is stored, let me know if this has helped at all
   
I don't think I'll be able to solve the audio issues fully, I can only reproduce sporadically when my phone is under load.

**Reply by sonht-0555** [reply]
@thenick775 
I tested it, it works very well.  The heating phenomenon no longer occurs.  It's almost the best version.
 - The sound has improved significantly, I tried plugging in headphones to test, the sound is almost 90-95/100% compared to running on the normal model
 - The image is no longer too jerky, I rate it 97/100%.  But running at 4X, I still clearly feel the frame jerking around (70/100%)
 - The phenomenon of stalling after running another heavy application is no longer there (tell me why and how you fix it?)

**Reply by sonht-0555** [reply]
https://gba.nicholas-vancise.dev/ I also tried testing here, you pushed these new things to product ah

**Reply by thenick775** [reply]
Yes I'm crafting the release now, as I'm happy with its current state 😎 

I'll write you up a details explanation here shortly about how I handled the backgrounding, and what I'll do in the future!

Great to hear that this fixed a bunch of those problems! I still have some improvements yet to come, so I hope to progressively make those smaller improvements as things shake out.

Thanks for all your engagement and thoughtful comments @sonht0555 !

And feel free to open issues, discussions or make PRs if you have other requests!

**Reply by sonht-0555** [reply]
@thenick775 
Ay, wonderful!!!  If there is a new feature, if you need someone to test it, please call me.  Would love to try out new features on this core

**Reply by sonht-0555** [reply]
@thenick775 
I'm trying to add new cores to my emulator.  My emulator is stopping just like yours. Reading how to fix it, you use `Module.pauseGame` `Module.resumeGame` to stop and restart the emulator every time it goes to the system tray, right?  
Sorry for my limited understanding of TS. If possible, can you guide me in implementing the fix logic using JS?

**Reply by sonht-0555** [reply]
I also tried debugging on iOS, but there was no error but the screen still froze, I still don't understand why it happened :/

**Reply by sonht-0555** [reply]
I have fixed it 
```
const handleVisibilityChange = () => {
    if (document.visibilityState === 'hidden') {
      Module.pauseGame();
      notiMessage("Game paused", 2000);
    } else {
      Module.resumeGame();
      notiMessage("Game resumed", 2000);
    }
  };
    document.addEventListener('visibilitychange', handleVisibilityChange);
    
    handleVisibilityChange();
```

**Reply by thenick775** [reply]
Yep that's exactly it!

The emulator now runs on a time delta, and it has no way to know if the page was backgrounded (i.e. not visible anymore).

Pausing the emulator using the visibility change event solves this by forcing the emulator to run the first frame instead of try and use the time delta to compute how many frames it should be running in a single go.

There are some safeties I can put in place in the core so that this doesn't hijack the main thread when it has too many frames to render, and I plan to hash that out sometime soon 😎

**Reply by sonht-0555** [reply]
@thenick775 
When you have free time, try emulating with core NBA. It's actually suitable for mobile devices, and it really runs well in the 30fps environment of the save model. The sound and frame are no longer jerky along with the popping sound like mgba 

https://nealmm.github.io

**Reply by thenick775** [reply]
How do you integrate with this core?

What is its contract in terms of functionality if you know?

I'm curious if it has full feature parity with mGBA.

**Reply by sonht-0555** [reply]
@thenick775 
https://nealmm.github.io/web-gba-emu/
This is his emulator. Its core is NanoBoyAdvance.
https://github.com/nba-emu/NanoBoyAdvance

- Its features are as rich as mGBA, but it has all the necessary modules we want. You can visit the link above to see what the NBA has to offer. It came out after mGBA so the community knows less about it, but its strengths are that it supports a lot of hardware, is light and fast. It simulates sound and images closest to reality (much better than mGBA, I don't rate mGBA with sound on iOS devices highly). SkyEmu also inherits this core, it carries NBA strengths such as color correction and LCD ghosting. You can see how SkyEmu displays beautiful colors, NBA does the same. 

- @Nealmm used Ems to compile. You can read the src/main.cpp file to see how he compiled Wasm.  My knowledge of Ems is quite limited, I tried creating an additional save and load state module. But not successful because of that. 

- Simply Fork it, create a Wasm branch and do your work like you did with mGBA (mGBA and NBA are both free open source emulators, be more careful and tag the author). fake into your RP like @Nealmm. I don't think there are any important legal constraints here. If you plan to build gbajs3 for the purpose of charging later, you should probably inform the author If you are worried about Nintendo suing, rest assured, NBA is just like mGBA or SkyEmu. They have enough reasons to overcome this giant.

**Reply by thenick775** [reply]
This repo will always remain public.

I'll take a look into this as I free up!

When you say your experiencing issues with iOS sound on my core, is that only in low power mode?

**Reply by sonht-0555** [reply]
With your knowledge about msc, I think it's not difficult for you to write full modules for it. Currently @Nealmm has only written basic modules. This core is really great. What @Nealmm gets out of it is so little, but its response on mobile devices is truly surprising. We don't need to correct or fix fps errors like mGBA 😂

**Reply by thenick775** [reply]
Feel free to write the glue code mirroring the wasm contract with mGBA and submit a PR if you get to this before I do!

**Reply by sonht-0555** [reply]
@thenick775 

What is glue code mirroring? i don't know about it :/. The above is the content I read on groups and discord. I don't know how to help you

**Reply by sonht-0555** [reply]
@thenick775 

accuracy, sound error and frame stutter at 30fps. I'm a perfectionist so I noticed that although it's not unpleasant if you get used to it. Even when running at 60fps, the sound is not as clear as NBA

**Reply by sonht-0555** [reply]
@thenick775 

I understand what you mean, my glue code is from your latest commit on gbajs3

**Reply by thenick775** [reply]
By glue code I mean something like [this](https://github.com/thenick775/mgba/blob/feature/wasm/src/platform/wasm/pre.js) in the pre.js files that allows for interactions from a frontend.

If we can mirror the features of the current [contract](https://github.com/thenick775/mgba/blob/feature/wasm/src/platform/wasm/mgba.d.ts) if they are already supported in NBA than its a lot easier to drop into place.

The glue code is what allows access from javascript, but also is what I'm using to define a standard set of functions I wish to support.

**Reply by sonht-0555** [reply]
@thenick775 
https://github.com/nealmm/web-gba-emu/blob/main/public/pre.js

you mean this one. it's in @nealmm's emulator. As far as I can see, @nealmm only took the necessary features of the NBA. They still have many features that need to be further exploited such as save-load state...

**Comment by thenick775** [comment, marked as answer]
Fixed in #120 🚀
